src/util/Thread.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "util/Thread.h"
+#include "util/Logging.h"
+
+#ifdef _WIN32
+#else
+#include <pthread.h>
+#endif
+
+namespace stellar
+{
+
+#ifdef _WIN32
+
+void
+runWithLowPriority(std::thread& thread)
+{
+    auto native = thread.native_handle();
+    auto ret = SetThreadPriority(native, THREAD_PRIORITY_BELOW_NORMAL);
+
+    if (!ret)
+    {
+        CLOG(DEBUG, "Fs") << "Unable to set priority for thread: " << ret;
+    }
+}
+
+#else
+
+void
+runWithLowPriority(std::thread& thread)
+{
+    auto native = thread.native_handle();
+    int policy;
+    sched_param param;
+
+    auto ret = pthread_getschedparam(native, &policy, &param);
```

Comment:
I am not sure I understand why this is "low priority": according to https://linux.die.net/man/2/sched_setscheduler `SCHED_BATCH` doesn't change priority only the policy when switching tasks.

Wasn't `pthread_setschedprio` enough (and maybe more portable)?

Reply:
(portability here in context of other POSIX systems like BSD/MacOS)

Reply:
According to http://man7.org/linux/man-pages/man7/sched.7.html:
- it is not possible to set priority != 0 for threads with policy of `SCHED_OTHER`, `SCHED_IDLE`, `SCHED_BATCH`
- `SCHED_OTHER` is default for threads on linux
- all other policies (`SCHED_DEADLINE`, `SCHED_RR`, `SCHED_FIFO`) are real-time policies, which have higher priority to all other threads, so priority value can only be used to further increase that priority, not to decrease it

So for changing thread priority we need to change both policy and priorty, it can be done by `pthread_setchedparam`.

Other way to reduce thread priority would be to `nice` it, but `nice` is per-process only according to POSIX. In Linux is can be made per-thread, but that is non portable.

So it only leaves us with two choices: `SCHED_BATCH` and `SCHED_IDLE`. `SCHED_IDLE` is described as being extremely low-priority, and `SCHED_BATCH` as suitable for long-running tasks. As our worker thread is mostly used for merging buckets, I've chosen `SCHED_BATCH` as appropriate value.

Reply:
read http://www.icir.org/gregor/tools/pthread-scheduling.html (and https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_setschedparam.3.html for MacOS ).

From what I get, `SCHED_FIFO` or `SCHED_RR` are the only "portable" ones. Even though they are global scheduling policies (so maybe need root?), to use them in this context would be to set a lower priority (ie, something between `sched_get_priority_min` and whatever `sched_getparam` returns).

If you want to use Linux specific stuff, we can just use `nice` ( http://man7.org/linux/man-pages/man2/nice.2.html ) that works on threads like you noticed.

Reply:
I've decided to use nice (this requires a bit of API changes, as it works on current thread).

Using FIFO or RR is not an option, because:
- it requires special permissions (by default root is needed)
- it only allows to increase priority (sched_getparam for main thread returns 0, sched_get_priority_min/max are 1-99 for FIFO and RR)